### PR TITLE
Added coloring for classes

### PR DESCRIPTION
**Pull request by jessica0801**
### Summary 

This pull request adds the dashboard implementations and implements the new designs with MaterialUI for when there are students to populate the class groupings with. It includes a class card for every class a student requests and colors the  students label for each card with the appropriate color. When there is more than 1 student who submits a form for a class, a class is matchable. In this case, the number of students label is green and when it is 1, it is yellow.

New features:

- the dashboard now calls the backend for the course info
- the CourseInfo type has been altered to fit the data from the json files received
- the group card and group files have been altered to retrieve and show the relevant information for grouping (ex. new students for a class and number of groups that have been formed)
- text and button components on the GroupCard now use materialUI.

Fixes:

- fixed the course info format to match the data fields of the json file received from the backend
- fixed the data parsing in dashboard to get the correct information from the function call

This pull request is the first step towards implementing the populated dashboard.

Files changed:

- [x] groupsicon.svg
- [x] newlymatchable.svg
- [x] plusicon.svg
- [x] Core/Constants/Theme.ts
- [x] GroupCard.tsx
- [x] GroupCard.style.tsx
- [x] Groups.tsx
- [x] GroupCard.style.tsx
- [x] warning.svg
- [x] Dashboard.tsx 
- [x] Dashboard.style.tsx 

### Test Plan 

I mainly tested this implementation by using Insomnia to retrieve the test data. I then inspected if the correct information was being displayed on the dashboard by comparing the displayed data to that in the json file. 

Screen Shot 2022-03-16 at 6 08 59 PM

### Notes 

The figma implements pagination for the classes on top of making the  content scrollable. However, the current implementation does not include this. The classes are scrollable in their container.
### Breaking Changes 
N/A

### CODE REVIEW

File: frontend/src/modules/Dashboard/Styles/GroupCard.style.tsx
Comment (mluo24) [inline frontend/src/modules/Dashboard/Styles/GroupCard.style.tsx]: I think here can be moved to one of the "Types" files, likely for "frontend/src/modules/Core/Types/StyledComponent.ts"

File: frontend/src/modules/Dashboard/Styles/GroupCard.style.tsx
Comment (mluo24) [inline frontend/src/modules/Dashboard/Styles/GroupCard.style.tsx]: One thing that I have issue with here is that you're making entirely new components for what seem to be trivial style changes (either one change or none at all), and with MUI v5, I believe the customization would be a lot easier with simple, one-off changes. You don't have MUI v5 in this branch, so for when you do, I will leave this here:
https://mui.com/system/the-sx-prop/
https://mui.com/customization/how-to-customize/

File: frontend/src/modules/Dashboard/Components/GroupCard.tsx
Comment (mluo24) [inline frontend/src/modules/Dashboard/Components/GroupCard.tsx]: Is there a reason you're using this sort of syntax? They're both numbers, so why can't you just pass them in to the function? like so:
`const styleArray = getColor(newStudents, groupsFormed)`

File: frontend/src/modules/Dashboard/Components/GroupCard.tsx
Comment (mluo24) [inline frontend/src/modules/Dashboard/Components/GroupCard.tsx:43]: Neat solution!

File: frontend/src/modules/Dashboard/Components/GroupCard.tsx
Comment (mluo24) [inline frontend/src/modules/Dashboard/Components/GroupCard.tsx]: Spelling error?

Comment (mluo24) [reply frontend/src/modules/Dashboard/Components/GroupCard.tsx]: Also it would be neat if you changed it to "student" or "students" based on how many students there are.

File: frontend/src/modules/Dashboard/Components/GroupCard.tsx
Comment (mluo24) [inline frontend/src/modules/Dashboard/Components/GroupCard.tsx]: Same here, changing the pluralization of groups accordingly!

File: frontend/src/modules/Dashboard/Components/GroupCard.tsx
Comment (mluo24) [inline frontend/src/modules/Dashboard/Components/GroupCard.tsx:98]: Does this button do anything yet?

**Review by mluo24 — CHANGES_REQUESTED** [review]
Good job on this feature! Your provided screenshots look really nice! You're doing really well with passing information from the request down to the props. I loved the way you rendered the colors and the information to present to the user.

I don't want to explicitly approve this yet because I just merged in the changes to upgrade everything to MUI v5, so there's a chance that if you merge in, there could be slight conflicts. I advise that you pull from main and resolve conflicts and make sure everything is running properly before you merge. I haven't actually added any component library changes yet so I can't tell you what to do with that yet, but the most important part is that your code works with MUI v5.

File: frontend/src/modules/Dashboard/Components/GroupCard.tsx
Comment (mluo24) [reply frontend/src/modules/Dashboard/Components/GroupCard.tsx]: Also I just noticed some of the "Groups Formed" don't have numbers beside them. Is this intentional due to some lack of information somewhere?

File: frontend/src/modules/Dashboard/Components/GroupCard.tsx
Comment (Coreball) [reply frontend/src/modules/Dashboard/Components/GroupCard.tsx:43]: I agree

File: frontend/src/modules/Dashboard/Components/GroupCard.tsx
Comment (Coreball) [reply frontend/src/modules/Dashboard/Components/GroupCard.tsx]: Pluralization is neat

File: frontend/src/modules/Dashboard/Components/GroupCard.tsx
Comment (Coreball) [reply frontend/src/modules/Dashboard/Components/GroupCard.tsx]: I think this is due to the Firebase default collection not having a field called `lastGroupNumber` for all documents. TypeScript's not gonna save us here because the data is only fetched at runtime, so the field can be missing even though the interface says it should exist. We should clean up the Firebase at some point and/or stop separating `courses` and `courses_test` collections because everything is in test mode right now.

File: frontend/src/modules/Dashboard/Components/GroupCard.tsx
Comment (Coreball) [reply frontend/src/modules/Dashboard/Components/GroupCard.tsx:98]: Not yet, but it should later lead to something of the form `/edit/350025`

File: frontend/src/modules/Dashboard/Styles/GroupCard.style.tsx
Comment (Coreball) [reply frontend/src/modules/Dashboard/Styles/GroupCard.style.tsx]: Yes once you merge in the changes from the MUI v5 migration, small style changes like this should be a lot simpler!

File: frontend/src/modules/Dashboard/Styles/GroupCard.style.tsx
Comment (Coreball) [inline frontend/src/modules/Dashboard/Styles/GroupCard.style.tsx]: Interesting, I haven't actually used Styled Components this way yet

**Review by Coreball — COMMENTED** [review]
Good job making progress on the LSC dashboard! Distinguishing between different types of matchable or unmatchable classes will greatly boost usability for the app. I believe there are some higher-contrast dashboard updates in the design pipeline, but we can figure those out next semester. For now, merge master into this branch to resolve conflicts and make sure everything still works with the MUI v5 migration that @mluo24 performed in #15. Great job!

File: frontend/src/modules/Dashboard/Styles/GroupCard.style.tsx
Comment (mluo24) [reply frontend/src/modules/Dashboard/Styles/GroupCard.style.tsx]: This is giving d3 data joins lol

File: frontend/src/modules/Dashboard/Components/GroupCard.tsx
Comment (mluo24) [inline frontend/src/modules/Dashboard/Components/GroupCard.tsx]: I love the attention to detail, but I believe the plural can be simplified to pretty much one line, and you don't need this duplication. Only do the conditional for the s, not the whole component.

File: frontend/src/modules/Dashboard/Components/GroupCard.tsx
Comment (mluo24) [inline frontend/src/modules/Dashboard/Components/GroupCard.tsx]: Same here for the plural.

File: frontend/src/modules/Dashboard/Styles/GroupCard.style.tsx
Comment (mluo24) [reply frontend/src/modules/Dashboard/Styles/GroupCard.style.tsx]: Remember to fix this part! especially the empty one

File: frontend/src/modules/Dashboard/Styles/GroupCard.style.tsx
Comment (mluo24) [inline frontend/src/modules/Dashboard/Styles/GroupCard.style.tsx]: Delete this comment.

File: frontend/src/modules/Dashboard/Components/GroupCard.tsx
Comment (mluo24) [inline frontend/src/modules/Dashboard/Components/GroupCard.tsx]: because the indexing can be sort of unclear, maybe use some sort of map with keys instead of just the array indexing? I can't parse what this means immediately. I get that it's from the arrays you return in line 43 though.

**Review by mluo24 — APPROVED** [review]
Good job on completing this task! The dashboard visuals are definitely an important step for displaying some organization to the admin, and your solution is pretty elegant! I recommend fixing some of the ideas I mentioned, but I'm approved this pull request because everything looks good!

File: frontend/src/modules/Dashboard/Styles/Dashboard.style.tsx
Comment (Coreball) [inline frontend/src/modules/Dashboard/Styles/Dashboard.style.tsx:11]: These lines can be removed now, judging from the designs

File: frontend/src/modules/Dashboard/Components/GroupCard.tsx
Comment (Coreball) [inline frontend/src/modules/Dashboard/Components/GroupCard.tsx]: The only difference between these is the "s" on "Groups"? We can shorten the code by doing something like `{groupsFormed === 1 ? 'Group' : 'Groups'}` as suggested by @mluo24 earlier. Similar for the new students part above

File: frontend/src/modules/Dashboard/Components/GroupCard.tsx
Comment (Coreball) [inline frontend/src/modules/Dashboard/Components/GroupCard.tsx]: I prefer how you did it for `<StyledNewlyMatchable />`, something like `<StyledWarningIcon />`. Same for StyledPlusIcon and StyledGroupsIcon. Also does `width: 1` in the CSS do anything? If not, you should remove it because it can be confusing

File: frontend/src/modules/Dashboard/Components/GroupCard.tsx
Comment (Coreball) [inline frontend/src/modules/Dashboard/Components/GroupCard.tsx]: I don't see `styleArray[1]` being used anywhere in the latest version of this PR. Suggest removing the middle array element. Also `styleArray[2]` is only being used in one place

File: frontend/src/modules/Dashboard/Components/GroupCard.tsx
Comment (Coreball) [inline frontend/src/modules/Dashboard/Components/GroupCard.tsx]: What does `justifyContent: ''` do? Is it necessary?

**Review by Coreball — APPROVED** [review]
Good work! I am excited to get this merged in very soon. I am approving this for now, but please make the few small changes I suggested in my PR comments.

I also made a new class "CS 4780" on Firestore where there are 0 unmatched students. I don't think this PR adheres exactly to the designs for that because I believe the designs use total number of students in the class in that situation? Leave this to a future pull request though, just get this merged in after making the small changes mentioned above.

Excited!

**Comment by gracshan** [comment]
Great work in implementing the new design! I especially like how you handled passing down the information from your request down through the rest of the code. One thing I might note is how you are using singular versus plural words in the variable names (students, groups). As long as you are consistent with these, I think you're good! 
Overall, good job!

**Pull request merged**
